### PR TITLE
Support chained controllers

### PR DESCRIPTION
**Pull request by pac48**
This PR adds support for ros2_control's chained controllers. Only simple chains are supported currently. That is, controller can only chain to at most one other controller. I also added logic to start and stop the chained controller in the correct order.

I created a demo repository under ros2_control with that uses the chnages made in this PR. Here is a link to the [README](https://github.com/pac48/ros2_control_demos/blob/add-admittance-controller/using_admittance_controller.md)

### CODE REVIEW

**Comment by v4hn** [comment]
> Only simple chains are supported currently. That is, controller can only chain to at most one other controller

Why? It does not sound like it would be too difficult to extend it?
That being said I do wonder whether it isn't ros_control's responsibility to tell you which hardware it is responsible to on top of which other controller it feeds into...

**Comment by pac48** [comment]
> > Only simple chains are supported currently. That is, controller can only chain to at most one other controller
> 
> Why? It does not sound like it would be too difficult to extend it? That being said I do wonder whether it isn't ros_control's responsibility to tell you which hardware it is responsible to on top of which other controller it feeds into...

The response from the ros2_control list controllers service does indicate which controllers are chained to which. The logic in this PR is really about taking that information and converting it to a format that MoveIt understands. 

Regarding chaining multiple controllers, MoveIt's controller manager plugin assumes there exists a direct mapping from controllers to controllable interfaces. If the controllers are chained with only one connection, then it is straight forward to propagate the controllable interfaces to the beginning of the chain, which ultimately controls those interfaces. If branching occurs in the chain, then propagating the  interfaces is ambiguous.

Lastly, I believe the ros2_control maintainers are working towards a simplified service interface to handle chained controllers. When that happens, handling the general case will be more straightforward.

**Comment by v4hn** [comment]
I'll leave the actual review here to people who work with ros2_control.

**Comment by AndyZe** [comment]
@pac48 it would help a lot if you could add test instructions to the PR description. I think one of your ros2_control PR's added a 6-dof example... maybe we can test this PR with that.

File: moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp
Comment (AndyZe) [inline moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp:113]: Does this only apply to chained controllers? May want to rename it `chained_dependencies_map_` or something similar, if so. And add comments.

```suggestion
  // Chained controllers have dependencies (other controllers which must be running)
  std::unordered_map<std::string /* controller name */, std::vector<std::string> /* dependencies */> dependency_map_;
```

File: moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp
Comment (AndyZe) [inline moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp:467]:
```suggestion
    for (const auto& c : result->controller)
```

File: moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp
Comment (AndyZe) [inline moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp:462]: Prob should make this a unique_ptr for safety's sake.
```suggestion
    std::unordered_map<std::string, std::unique_ptr<controller_manager_msgs::msg::ControllerState>> controller_name_map;
```

File: moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp
Comment (pac48) [reply moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp:113]: It does only apply to chained controllers for now, so the name should be changed.

File: moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp
Comment (pac48) [reply moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp:383]: Unfortunately, the order in which the chained controllers is started matters. Dependencies must be started first, but they are processed last, so the order needs to be flipped to start everything in the correct order.

File: moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp
Comment (tylerjw) [inline moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp:465]:
```suggestion
    for (size_t i = 0; i < result->controller.size(); ++i)
```

**Review by tylerjw — APPROVED** [review]
One tiny nit. Otherwise, this looks good.

File: moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp
Comment (AndyZe) [inline moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp:383]:
```suggestion
    // activation dependencies must be started first, but they are processed last, so the order needs to be flipped
```

File: moveit_plugins/moveit_ros_control_interface/src/empty_controller_plugin.cpp
Comment (AndyZe) [inline moveit_plugins/moveit_ros_control_interface/src/empty_controller_plugin.cpp:44]:
```suggestion
 * \brief This class allows MoveIt's controller manager to start and stop a controller that does not have a
```

File: moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/empty_controller_handle.h
Comment (AndyZe) [inline moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/empty_controller_handle.h:46]:
```suggestion
 * An interface for controllers that have no handle, e.g. chained controllers like an Admittance controller
```

**Review by AndyZe — APPROVED** [review]
I've tested it now, and also tested that it doesn't break the regular Joint Trajectory Controller. Once you rebase, we should be clear to merge it.   :)

**Pull request merged**
